src/misc/bleeding.ts

```diff
@@ -0,0 +1,204 @@
+  ) {
+    this.entity = entity;
+    this.amount = amount;
+
+    const handle = setInterval(() => {
+      try {
+        if (!this.tick()) throw undefined;
+
+        this.time += TIME_STEP;
+      } catch (error) {
+        clearInterval(handle);
+        callback(error);
+      }
+    }, TIMER_DELAY);
+  }
+
+  /**
+   * Handle timer cycle
+   */
+  tick() {
+    // Spawn blood at exponential rate
+    const chance = this.amount * Math.pow(COEFFICIENT, this.time);
+
+    this.spawnBlood(this.entity.location, 1.0, chance);
+
+    // Stop task if entity either stops bleeding or dies
+    return chance > 0.1 && !this.entity.isDead();
+  }
+
+  /**
+   * Increases bleeding amount and duration
+   * @param amount from the range of 0.0 to 1.0
+   */
+  extend(amount: number) {
+    // Cap value to 1.0
+    this.amount = Math.max(0.0, Math.min(this.amount + amount, 1.0));
+
+    this.time = 0.0;
+  }
+
+  /**
+   * Spawn blood particles at location.
+   *
+   * * Note that this method relies on Paper API
+   */
+  spawnBloodParticles(location: Location, count: number) {
+    new ParticleBuilder(Particle.BLOCK_DUST)
+      .location(location)
+      .data(PARTICLE_DATA)
+      .count(count)
+      .spawn();
+  }
+
+  /**
+   * Spawn blood puddles and splashes
+   * @param radius blood spawn radius from the entity outwards
+   * @param probability
+   */
+  spawnBlood(location: Location, radius: number, probability: number) {
+    const origin = location.block;
+
+    for (let x = -radius; x <= radius; x++) {
+      for (let z = -radius; z <= radius; z++) {
+        if (Math.random() > probability) continue;
+
+        // Compare against squared radius to check that point is inside circle
+        if (x * x + z * z > radius * radius) continue;
+
+        // Splatter blood
+        this.spawnBloodParticles(
+          location.clone().add(0.0, 0.8, 0.0),
+          PARTICLE_AMOUNT,
+        );
+
+        const block = origin.getRelative(x, 0.0, z);
+
+        // Check that block and location is valid
+        if (
+          !block.type.isAir() ||
+          !block.getRelative(BlockFace.UP).type.isAir() ||
+          !block.getRelative(BlockFace.DOWN).type.isOccluding()
+        )
+          return;
+
+        // Form a puddle
+        block.type = BLOOD_MATERIAL;
+        const data = block.blockData as Waterlogged;
+        data.setWaterlogged(false);
+        block.blockData = data;
+      }
+    }
+  }
+}
+
+// Active tasks by entity id
+const tasks = new Map<number, BleedTask>();
+
+/**
+ * Sets entity bleeding or extends bleeding duration and amount
+ * @param entity
+ * @param amount from the range of 0.0 to 1.0
+ */
+function setBleeding(entity: Entity, amount: number) {
+  let task = tasks.get(entity.entityId);
+
+  if (task) {
+    // Increase bleeding amount and duration
+    task.extend(amount);
+
+    return;
+  }
+
+  // Create a new task
+  task = new BleedTask(entity, amount, (error: Error) => {
+    if (error) {
```

Comment:
Note that catching error in JS side and logging it to console will cause the stack trace to be lost. This seems to be a GraalJS limitation.